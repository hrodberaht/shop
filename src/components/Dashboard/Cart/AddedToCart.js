import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class AddedToCart extends Component {
  render() {
    const {
      product: {
        name, price, pcsOrder, totalPrice,
      },
      show,
      redirect,
    } = this.props;
    return (
      <div className="cart-popup">
        <div className="cart-popup__inside">
          <h4>Product added to cart:</h4>
          <p>
            <span>Name:</span>
            {name}
          </p>
          <p>
            <span>Price:</span>
            {price}
          </p>
          <p>
            <span>Pcs:</span>
            {pcsOrder}
          </p>
          <p>
            <span>Tota price:</span>
            {totalPrice}
          </p>
          <div className="cart-popup__buttons">
            <button id="more" className="btn btn-primary" type="button" onClick={show}>
              More
            </button>
            <button id="cart" className="btn btn-primary" type="button" onClick={redirect}>
              Cart
            </button>
          </div>
        </div>
      </div>
    );
  }
}

AddedToCart.propTypes = {
  product: PropTypes.shape({
    name: PropTypes.string,
    price: PropTypes.number,
    pcsOrder: PropTypes.number,
    totalPrice: PropTypes.number,
  }).isRequired,
  show: PropTypes.func.isRequired,
  redirect: PropTypes.func.isRequired,
};
