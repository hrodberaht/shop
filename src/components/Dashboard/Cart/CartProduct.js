import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class CartProduct extends Component {
  render() {
    const {
      product: {
        name, price, pcsOrder, totalPrice,
      },
    } = this.props;
    return (
      <tr>
        <td>{name}</td>
        <td>{price}</td>
        <td>{pcsOrder}</td>
        <td>
          <span>$</span>
          {totalPrice}
        </td>
      </tr>
    );
  }
}

CartProduct.propTypes = {
  product: PropTypes.shape({
    name: PropTypes.string,
    price: PropTypes.number,
    pcsOrder: PropTypes.number,
    totalPrice: PropTypes.number,
  }).isRequired,
};
