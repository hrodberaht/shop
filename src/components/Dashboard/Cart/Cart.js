import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import moment from 'moment';
import { getProductsInCart, getOrderPositionIds } from '../../../store/cart/selectors';
import CartProduct from './CartProduct';
import { clearCart, removeFromCart } from '../../../store/cart/actionCreators';
import { addOrderToDB } from '../../../store/orders/actionCreators';
import {
  getAuthUserId,
  getAuthToken,
  getAuthPerson,
  getAuthCompanyId,
} from '../../../store/authenticate/selectors';

export class Cart extends Component {
  sumaryPrice = () => {
    const {
      cart: { list, byId },
    } = this.props;
    let sumary = 0;
    list.map((item) => {
      sumary += byId[item].totalPrice;
      return sumary;
    });
    return sumary;
  };

  remove = product => this.props.removeFrom(product);

  handleClick = () => {
    const {
      emptyCart,
      createOrder,
      userId,
      person,
      token,
      companyId,
      orderPositionIds,
    } = this.props;
    const order = {
      userId,
      person,
      totalPrice: +this.sumaryPrice(),
      status: 'in-progress',
      date: moment().format('MMMM Do YYYY, h:mm:ss a'),
      companyId,
      orderPositionIds,
    };
    createOrder(order, token);
    emptyCart();
    this.props.history.push('/orders');
  };

  render() {
    const {
      cart: { list, byId },
    } = this.props;
    return (
      <div className="cart">
        <h3>Products in your cart:</h3>
        <table>
          <tbody className="cart-produts__item">
            <tr>
              <th>Name:</th>
              <th>Price:</th>
              <th>Pcs:</th>
              <th>Total price:</th>
              <th>Delete:</th>
            </tr>
            {list.map(item => (
              <CartProduct key={byId[item].id} product={byId[item]} remove={this.remove} />
            ))}
          </tbody>
        </table>
        <p className="cart__sumary">
          <span>$</span>
          {this.sumaryPrice()}
        </p>
        <button
          className="btn btn-primary"
          type="button"
          onClick={this.handleClick}
          disabled={list.length === 0}
        >
          Buy
        </button>
      </div>
    );
  }
}

withRouter(Cart);
const mapStateToProps = state => ({
  cart: getProductsInCart(state),
  userId: getAuthUserId(state),
  token: getAuthToken(state),
  person: getAuthPerson(state),
  companyId: getAuthCompanyId(state),
  orderPositionIds: getOrderPositionIds(state),
});

export default connect(
  mapStateToProps,
  {
    emptyCart: clearCart,
    createOrder: addOrderToDB,
    removeFrom: removeFromCart,
  },
)(Cart);

Cart.propTypes = {
  cart: PropTypes.shape({
    list: PropTypes.array,
    byId: PropTypes.object,
  }).isRequired,
  emptyCart: PropTypes.func.isRequired,
  createOrder: PropTypes.func.isRequired,
  removeFrom: PropTypes.func.isRequired,
  companyId: PropTypes.string,
  userId: PropTypes.string.isRequired,
  person: PropTypes.string.isRequired,
  token: PropTypes.string.isRequired,
  orderPositionIds: PropTypes.arrayOf(PropTypes.arrayOf).isRequired,
};

Cart.defaultProps = {
  companyId: null,
};
