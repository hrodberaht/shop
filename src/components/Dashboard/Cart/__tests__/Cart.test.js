import { Cart } from '../Cart';

const defaultProps = {
  products: [
    {
      id: '1',
      name: 'fx',
      price: 400,
      pcsOrder: 2,
      totalPrice: 800,
    },
    {
      id: '2',
      name: 'fx2',
      price: 600,
      pcsOrder: 8,
      totalPrice: 4800,
    },
  ],
  emptyCart: jest.fn(),
  createOrder: jest.fn(),
  history: [],
  removeFrom: jest.fn(),
  userId: '1234',
  person: 'Tom Johns',
  token: '1234',
  orderPositionIds: [],
};

const setup = buildSetup(Cart, defaultProps);

describe('<Cart />', () => {
  it('should render witout crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call empty cart after click', () => {
    const { wrapper } = setup();
    wrapper.find('button').simulate('click');
    setTimeout(() => {
      expect(wrapper.instance().props.emptyCart).toHaveBeenCalled();
    }, 1000);
  });

  it('return sumary price', () => {
    const { wrapper } = setup();
    wrapper.instance().sumaryPrice();
    expect(wrapper.instance().sumaryPrice()).toBe(5600);
  });
  it('remove should call removeFrom', () => {
    const { wrapper } = setup();
    wrapper.instance().remove();
    expect(wrapper.instance().props.removeFrom).toHaveBeenCalled();
  });
});
