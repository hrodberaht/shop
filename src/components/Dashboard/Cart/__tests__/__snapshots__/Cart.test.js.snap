// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> should render witout crash 1`] = `
<div
  className="cart"
>
  <h3>
    Products in your cart:
  </h3>
  <table>
    <tbody
      className="cart-produts__item"
    >
      <tr>
        <th>
          Name:
        </th>
        <th>
          Price:
        </th>
        <th>
          Pcs:
        </th>
        <th>
          Total price:
        </th>
        <th>
          Delete:
        </th>
      </tr>
      <CartProduct
        changeQuantity={[Function]}
        key="1234"
        product={
          Object {
            "id": "1234",
            "name": "fx",
            "pcsOrder": 2,
            "price": 400,
            "productId": "1",
            "totalPrice": 800,
          }
        }
        remove={[Function]}
      />
      <CartProduct
        changeQuantity={[Function]}
        key="5678"
        product={
          Object {
            "id": "5678",
            "name": "fx2",
            "pcsOrder": 8,
            "price": 600,
            "productId": "2",
            "totalPrice": 4800,
          }
        }
        remove={[Function]}
      />
    </tbody>
  </table>
  <div
    className="cart__sumary"
  >
    <p>
      $
      5600
    </p>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Buy
    </button>
  </div>
</div>
`;
